```
jects\tree-sitter-0.19.5\lib\include" "-I..\subprojects\libuv-v1.40.0\include" "/MT" "/nologo" "/showIncludes" "/W2" "/O2" "/Gw" "" "-DRZ_PLUGIN_INCORE=1" "/Fdlibrz/core\librz_core.a.p\cconfig.c.pdb" /Folibrz/core/librz_core.a.p/cconfig.c.obj "/c" ../librz/core/cconfig.c
cl : Command line warning D9024 : unrecognized source file type '', object file assumed
cl : Command line warning D9027 : source file '' ignored
../librz/core/cconfig.c(2196): error C2065: 'RZ_VIRT_TERM_MODE_DISABLED': undeclared identifier
[1187/1678] Compiling C object librz/core/librz_core.a.p/cdwarf.c.obj
cl : Command line warning D9024 : unrecognized source file type '', object file assumed
cl : Command line warning D9027 : source file '' ignored
[1188/1678] Compiling C object librz/core/librz_core.a.p/chash.c.obj
cl : Command line warning D9024 : unrecognized source file type '', object file assumed
cl : Command line warning D9027 : source file '' ignored
ninja: build stopped: subcommand failed.
Command exited with code 1
```

--- a/librz/cons/cons.c
+++ b/librz/cons/cons.c
@@ -140,7 +140,7 @@ static void __break_signal(int sig) {
 
 static inline void __cons_write_ll(const char *buf, int len) {
 #if __WINDOWS__
-	if (I.vtmode) {
+	if (I.vtmode != RZ_VIRT_TERM_MODE_DISABLE) {
 		rz_xwrite(I.fdout, buf, len);
 	} else {
 		if (I.fdout == 1) {
@@ -493,7 +493,7 @@ RZ_API bool rz_cons_enable_mouse(const bool enable) {
 		return I.mouse;
 	}
 #if __WINDOWS__
-	if (I.vtmode == 2) {
+	if (I.vtmode == RZ_VIRT_TERM_MODE_COMPLETE) {
 #endif
 		const char *click = enable
 			? "\x1b[?1000;1006;1015h"
@@ -589,10 +589,10 @@ RZ_API RzCons *rz_cons_new(void) {
 #if __WINDOWS__
 	I.old_cp = GetConsoleCP();
 	I.old_ocp = GetConsoleOutputCP();
-	I.vtmode = rz_cons_is_vtcompat();
+	I.vtmode = rz_cons_detect_vt_mode();
 	set_console_codepage_to_utf8();
 #else
-	I.vtmode = 2;
+	I.vtmode = RZ_VIRT_TERM_MODE_COMPLETE;
 #endif
 #if EMSCRIPTEN
 	/* do nothing here :? */
@@ -721,7 +721,7 @@ RZ_API void rz_cons_fill_line(void) {
 
 RZ_API void rz_cons_clear_line(int std_err) {
 #if __WINDOWS__
-	if (I.vtmode) {
+	if (I.vtmode != RZ_VIRT_TERM_MODE_DISABLE) {
 		fprintf(std_err ? stderr : stdout, "%s", RZ_CONS_CLEAR_LINE);
 	} else {
 		char white[1024];
@@ -1036,7 +1036,7 @@ RZ_API void rz_cons_visual_flush(void) {
 	if (!I.null) {
 /* TODO: this ifdef must go in the function body */
 #if __WINDOWS__
-		if (I.vtmode) {
+		if (I.vtmode != RZ_VIRT_TERM_MODE_DISABLE) {
 			rz_cons_visual_write(I.context->buffer);
 		} else {
 			rz_cons_w32_print(I.context->buffer, I.context->buffer_len, true);
@@ -1071,7 +1071,7 @@ RZ_API void rz_cons_print_fps(int col) {
 		col = 12;
 	}
 #ifdef __WINDOWS__
-	if (I.vtmode) {
+	if (I.vtmode != RZ_VIRT_TERM_MODE_DISABLE) {
 		eprintf("\x1b[0;%dH[%d FPS] \n", w - col, fps);
 	} else {
 		rz_cons_w32_gotoxy(2, w - col, 0);
@@ -1510,19 +1510,19 @@ RZ_API int rz_cons_get_size(int *rows) {
 }
 
 #if __WINDOWS__
-RZ_API int rz_cons_is_vtcompat(void) {
+RZ_API RzVirtTermMode rz_cons_detect_vt_mode(void) {
 	DWORD major;
 	DWORD minor;
 	DWORD release = 0;
 	char *wt_session = rz_sys_getenv("WT_SESSION");
 	if (wt_session) {
 		free(wt_session);
-		return 2;
+		return RZ_VIRT_TERM_MODE_COMPLETE;
 	}
 	char *alacritty = rz_sys_getenv("ALACRITTY_LOG");
 	if (alacritty) {
 		free(alacritty);
-		return 1;
+		return RZ_VIRT_TERM_MODE_OUTPUT_ONLY;
 	}
 	char *term = rz_sys_getenv("TERM");
 	if (term) {
@@ -1537,9 +1537,9 @@ RZ_API int rz_cons_is_vtcompat(void) {
 	char *ansicon = rz_sys_getenv("ANSICON");
 	if (ansicon) {
 		free(ansicon);
-		return 1;
+		return RZ_VIRT_TERM_MODE_OUTPUT_ONLY;
 	}
-	bool win_support = 0;
+	RzVirtTermMode win_support = RZ_VIRT_TERM_MODE_DISABLE;
 	RSysInfo *info = rz_sys_info();
 	if (info && info->version) {
 		char *dot = strtok(info->version, ".");
@@ -1550,7 +1550,7 @@ RZ_API int rz_cons_is_vtcompat(void) {
 			release = atoi(info->release);
 		}
 		if (major > 10 || (major == 10 && minor > 0) || (major == 10 && minor == 0 && release >= 1703)) {
-			win_support = 1;
+			win_support = RZ_VIRT_TERM_MODE_OUTPUT_ONLY;
 		}
 	}
 	rz_sys_info_free(info);
@@ -1560,7 +1560,7 @@ RZ_API int rz_cons_is_vtcompat(void) {
 
 RZ_API void rz_cons_show_cursor(int cursor) {
 #if __WINDOWS__
-	if (I.vtmode) {
+	if (I.vtmode != RZ_VIRT_TERM_MODE_DISABLE) {
 #endif
 		rz_xwrite(1, cursor ? "\x1b[?25h" : "\x1b[?25l", 6);
 #if __WINDOWS__
@@ -1664,7 +1664,7 @@ RZ_API bool rz_cons_set_cup(bool enable) {
 	}
 	fflush(stdout);
 #elif __WINDOWS__
-	if (I.vtmode) {
+	if (I.vtmode != RZ_VIRT_TERM_MODE_DISABLE) {
 		if (enable) {
 			const char *code = enable // xterm + xterm-color
 				? "\x1b[?1049h\x1b"
@@ -1968,7 +1968,7 @@ RZ_API void rz_cons_cmd_help(const char *help[], bool use_color) {
 }
 
 RZ_API void rz_cons_clear_buffer(void) {
-	if (I.vtmode) {
+	if (I.vtmode != RZ_VIRT_TERM_MODE_DISABLE) {
 		rz_xwrite(1, "\x1b"
 			     "c\x1b[3J",
 			6);

--- a/librz/cons/dietline.c
+++ b/librz/cons/dietline.c
@@ -224,7 +224,7 @@ static int rz_line_readchar_win(ut8 *s, int slen) { // this function handle the 
 	void *bed;
 
 	h = GetStdHandle(STD_INPUT_HANDLE);
-	DWORD new_mode = I.vtmode == 2 ? ENABLE_VIRTUAL_TERMINAL_INPUT : 0;
+	DWORD new_mode = I.vtmode == RZ_VIRT_TERM_MODE_COMPLETE ? ENABLE_VIRTUAL_TERMINAL_INPUT : 0;
 	GetConsoleMode(h, &mode);
 	SetConsoleMode(h, new_mode);
 	if (I.zerosep) {
@@ -1611,7 +1611,7 @@ RZ_API const char *rz_line_readline_cb(RzLineReadCallback cb, void *user) {
 			break;
 		case 27: // esc-5b-41-00-00 alt/meta key
 #if __WINDOWS__
-			if (I.vtmode != 2) {
+			if (I.vtmode != RZ_VIRT_TERM_MODE_COMPLETE) {
 				memmove(buf, buf + 1, strlen(buf));
 				if (!buf[0]) {
 					buf[0] = -1;
@@ -1673,7 +1673,7 @@ RZ_API const char *rz_line_readline_cb(RzLineReadCallback cb, void *user) {
 				}
 				break;
 			default:
-				if (I.vtmode == 2) {
+				if (I.vtmode == RZ_VIRT_TERM_MODE_COMPLETE) {
 					buf[1] = rz_cons_readchar_timeout(50);
 					if (buf[1] == -1) { // alt+e
 						rz_cons_break_pop();
@@ -1685,7 +1685,7 @@ RZ_API const char *rz_line_readline_cb(RzLineReadCallback cb, void *user) {
 					switch (buf[1]) {
 					case '3': // supr
 						__delete_next_char();
-						if (I.vtmode == 2) {
+						if (I.vtmode == RZ_VIRT_TERM_MODE_COMPLETE) {
 							buf[1] = rz_cons_readchar();
 							if (buf[1] == -1) {
 								rz_cons_break_pop();
@@ -1694,7 +1694,7 @@ RZ_API const char *rz_line_readline_cb(RzLineReadCallback cb, void *user) {
 						}
 						break;
 					case '5': // pag up
-						if (I.vtmode == 2) {
+						if (I.vtmode == RZ_VIRT_TERM_MODE_COMPLETE) {
 							buf[1] = rz_cons_readchar();
 						}
 						if (I.hud) {
@@ -1709,7 +1709,7 @@ RZ_API const char *rz_line_readline_cb(RzLineReadCallback cb, void *user) {
 						}
 						break;
 					case '6': // pag down
-						if (I.vtmode == 2) {
+						if (I.vtmode == RZ_VIRT_TERM_MODE_COMPLETE) {
 							buf[1] = rz_cons_readchar();
 						}
 						if (I.hud) {
@@ -1784,7 +1784,7 @@ RZ_API const char *rz_line_readline_cb(RzLineReadCallback cb, void *user) {
 						__move_cursor_left();
 						break;
 					case 0x31: // control + arrow
-						if (I.vtmode == 2) {
+						if (I.vtmode == RZ_VIRT_TERM_MODE_COMPLETE) {
 							ch = rz_cons_readchar();
 							if (ch == 0x7e) { // HOME in screen/tmux
 								// corresponding END is 0x34 below (the 0x7e is ignored there)
@@ -1833,7 +1833,7 @@ RZ_API const char *rz_line_readline_cb(RzLineReadCallback cb, void *user) {
 							}
 							break;
 						default:
-							if (I.vtmode == 2) {
+							if (I.vtmode == RZ_VIRT_TERM_MODE_COMPLETE) {
 								if (I.cb_fkey) {
 									I.cb_fkey(I.user, fkey);
 								}

--- a/librz/cons/input.c
+++ b/librz/cons/input.c
@@ -99,7 +99,7 @@ static bool is_arrow;
 
 RZ_API int rz_cons_arrow_to_hjkl(int ch) {
 #if __WINDOWS__
-	if (I->vtmode != 2) {
+	if (I->vtmode != RZ_VIRT_TERM_MODE_COMPLETE) {
 		if (is_arrow) {
 			switch (ch) {
 			case VK_DOWN: // key down
@@ -442,7 +442,7 @@ static int __cons_readchar_w32(ut32 usec) {
 	h = GetStdHandle(STD_INPUT_HANDLE);
 	GetConsoleMode(h, &mode);
 	DWORD newmode = ENABLE_WINDOW_INPUT;
-	if (I->vtmode == 2) {
+	if (I->vtmode == RZ_VIRT_TERM_MODE_COMPLETE) {
 		newmode |= ENABLE_VIRTUAL_TERMINAL_INPUT;
 	}
 	newmode |= mode;

--- a/librz/cons/line.c
+++ b/librz/cons/line.c
@@ -26,9 +26,9 @@ RZ_API RzLine *rz_line_new(void) {
 	I.kill_ring = rz_list_newf(NULL);
 	I.kill_ring_ptr = -1;
 #if __WINDOWS__
-	I.vtmode = rz_cons_is_vtcompat();
+	I.vtmode = rz_cons_detect_vt_mode();
 #else
-	I.vtmode = 2;
+	I.vtmode = RZ_VIRT_TERM_MODE_COMPLETE;
 #endif
 	if (!rz_line_dietline_init()) {
 		eprintf("error: rz_line_dietline_init\n");

--- a/librz/cons/output.c
+++ b/librz/cons/output.c
@@ -25,7 +25,7 @@ RZ_API void rz_cons_w32_clear(void) {
 	static CONSOLE_SCREEN_BUFFER_INFO csbi;
 	COORD startCoords;
 	DWORD dummy;
-	if (I->vtmode) {
+	if (I->vtmode != RZ_VIRT_TERM_MODE_DISABLE) {
 		rz_cons_strcat(Color_RESET RZ_CONS_CLEAR_SCREEN);
 		return;
 	}
@@ -55,7 +55,7 @@ RZ_API void rz_cons_w32_gotoxy(int fd, int x, int y) {
 	COORD coord;
 	coord.X = x;
 	coord.Y = y;
-	if (I->vtmode) {
+	if (I->vtmode != RZ_VIRT_TERM_MODE_DISABLE) {
 		rz_cons_printf("\x1b[%d;%dH", y, x);
 		return;
 	}
@@ -411,7 +411,7 @@ RZ_API int rz_cons_win_vhprintf(DWORD hdl, bool vmode, const char *fmt, va_list 
 	FILE *con = hdl == STD_OUTPUT_HANDLE ? stdout : stderr;
 	if (!strchr(fmt, '%')) {
 		size_t len = strlen(fmt);
-		if (I->vtmode) {
+		if (I->vtmode != RZ_VIRT_TERM_MODE_DISABLE) {
 			return fwrite(fmt, 1, len, con);
 		}
 		return rz_cons_w32_hprint(hdl, fmt, len, vmode);
@@ -422,7 +422,7 @@ RZ_API int rz_cons_win_vhprintf(DWORD hdl, bool vmode, const char *fmt, va_list 
 	char *buf = calloc(1, num_chars);
 	if (buf) {
 		(void)vsnprintf(buf, num_chars, fmt, ap);
-		if (I->vtmode) {
+		if (I->vtmode != RZ_VIRT_TERM_MODE_DISABLE) {
 			ret = fwrite(buf, 1, num_chars - 1, con);
 		} else {
 			ret = rz_cons_w32_hprint(hdl, buf, num_chars - 1, vmode);

--- a/librz/core/cconfig.c
+++ b/librz/core/cconfig.c
@@ -2,6 +2,7 @@
 // SPDX-License-Identifier: LGPL-3.0-only
 
 #include <rz_core.h>
+#include <rz_cons.h>
 
 #include "core_private.h"
 
@@ -2176,15 +2177,15 @@ static bool cb_scrhighlight(void *user, void *data) {
 static bool scr_vtmode(void *user, void *data) {
 	RzConfigNode *node = (RzConfigNode *)data;
 	if (rz_str_is_true(node->value)) {
-		node->i_value = 1;
+		node->i_value = RZ_VIRT_TERM_MODE_OUTPUT_ONLY;
 	}
-	node->i_value = node->i_value > 2 ? 2 : node->i_value;
+	node->i_value = node->i_value > RZ_VIRT_TERM_MODE_COMPLETE ? RZ_VIRT_TERM_MODE_COMPLETE : node->i_value;
 	rz_line_singleton()->vtmode = rz_cons_singleton()->vtmode = node->i_value;
 
 	DWORD mode;
 	HANDLE input = GetStdHandle(STD_INPUT_HANDLE);
 	GetConsoleMode(input, &mode);
-	if (node->i_value == 2) {
+	if (node->i_value == RZ_VIRT_TERM_MODE_COMPLETE) {
 		SetConsoleMode(input, mode & ENABLE_VIRTUAL_TERMINAL_INPUT);
 		rz_cons_singleton()->term_raw |= ENABLE_VIRTUAL_TERMINAL_INPUT;
 	} else {
@@ -2193,7 +2194,7 @@ static bool scr_vtmode(void *user, void *data) {
 	}
 	HANDLE streams[] = { GetStdHandle(STD_OUTPUT_HANDLE), GetStdHandle(STD_ERROR_HANDLE) };
 	int i;
-	if (node->i_value > 0) {
+	if (node->i_value > RZ_VIRT_TERM_MODE_DISABLE) {
 		for (i = 0; i < RZ_ARRAY_SIZE(streams); i++) {
 			GetConsoleMode(streams[i], &mode);
 			SetConsoleMode(streams[i],

--- a/librz/core/visual.c
+++ b/librz/core/visual.c
@@ -4036,7 +4036,7 @@ RZ_API int rz_core_visual(RzCore *core, const char *input) {
 		if (!skip) {
 			ch = rz_cons_readchar();
 
-			if (I->vtmode == 2 && !is_mintty(core->cons)) {
+			if (I->vtmode == RZ_VIRT_TERM_MODE_COMPLETE && !is_mintty(core->cons)) {
 				// Prevent runaway scrolling
 				if (IS_PRINTABLE(ch) || ch == '\t' || ch == '\n') {
 					flush_stdin();

--- a/librz/include/rz_cons.h
+++ b/librz/include/rz_cons.h
@@ -446,10 +446,18 @@ typedef void (*RzConsSleepEndCallback)(void *core, void *user);
 typedef void (*RzConsQueueTaskOneshot)(void *core, void *task, void *user);
 typedef void (*RzConsFunctionKey)(void *core, int fkey);
 
-typedef enum { COLOR_MODE_DISABLED = 0,
+typedef enum {
+	COLOR_MODE_DISABLED = 0,
 	COLOR_MODE_16,
 	COLOR_MODE_256,
-	COLOR_MODE_16M } RzConsColorMode;
+	COLOR_MODE_16M
+} RzConsColorMode;
+
+typedef enum {
+	RZ_VIRT_TERM_MODE_DISABLE = 0, ///< Windows only: Use console c api for everything (Windows <= 8)
+	RZ_VIRT_TERM_MODE_OUTPUT_ONLY, ///< Windows only: Use console c api for input, but output on VT (Windows >= 10)
+	RZ_VIRT_TERM_MODE_COMPLETE, ///< All the sequences goes through VT (Windows Terminal, mintty, all OSs)
+} RzVirtTermMode;
 
 typedef struct rz_cons_context_t {
 	RzConsGrep grep;
@@ -537,7 +545,7 @@ typedef struct rz_cons_t {
 	const char **vline;
 	int refcnt;
 	RZ_DEPRECATE bool newline;
-	int vtmode;
+	RzVirtTermMode vtmode;
 	bool flush;
 	bool use_utf8; // use utf8 features
 	bool use_utf8_curvy; // use utf8 curved corners
@@ -858,7 +866,7 @@ RZ_API int rz_cons_pipe_open(const char *file, int fdn, int append);
 RZ_API void rz_cons_pipe_close(int fd);
 
 #if __WINDOWS__
-RZ_API int rz_cons_is_vtcompat(void);
+RZ_API RzVirtTermMode rz_cons_detect_vt_mode(void);
 RZ_API void rz_cons_w32_clear(void);
 RZ_API void rz_cons_w32_gotoxy(int fd, int x, int y);
 RZ_API int rz_cons_w32_print(const char *ptr, int len, bool vmode);
@@ -1138,7 +1146,7 @@ struct rz_line_t {
 	RzLineHud *hud;
 	RzList *sdbshell_hist;
 	RzListIter *sdbshell_hist_iter;
-	int vtmode;
+	RzVirtTermMode vtmode;
 }; /* RzLine */
 
 #ifdef RZ_API

